src/xtgeo/xyz/_xyz_roxapi.py

```diff
@@ -305,237 +399,139 @@ def _roxapi_export_xyz_viafile(
 
 
 def _roxapi_export_xyz(
-    self, rox, name, category, stype, pfilter, realisation, attributes
-):  # pragma: no cover
+    self: xtgeo.Points | xtgeo.Polygons,
+    rox: RoxUtils,
+    name: str,
+    category: str | list[str] | None,
+    stype: STYPE,
+    pfilter: dict[str, list] | None,
+    realisation: int,
+    attributes: bool,
+) -> None:
     logger.warning("Realisation %s not in use", realisation)
 
-    proj = rox.project
-    if not _check_category_etc(proj, name, category, stype, realisation, mode="set"):
-        raise RuntimeError("Cannot access correct category or name in RMS")
+    if self.dataframe is None or self.dataframe.empty:
+        logger.warning("Empty dataframe! Skipping object update...")
+        return
 
-    roxxyz = _get_roxitem(self, proj, name, category, stype, mode="set")
+    is_polygons = isinstance(self, xtgeo.Polygons)
 
-    if self.dataframe is None or len(self.dataframe.index) == 0:
-        return
+    roxxyz = _get_roxitem(
+        rox, name, category, stype, mode="set", is_polygons=is_polygons
+    )
 
-    dfrcopy = self.dataframe.copy()
-    # apply pfilter if any
-    if pfilter:
-        for key, val in pfilter.items():
-            if key in dfrcopy.columns:
-                dfrcopy = dfrcopy.loc[dfrcopy[key].isin(val)]
-            else:
-                raise KeyError(
-                    f"The requested pfilter key {key} was not found in dataframe. "
-                    f"Valid keys are {dfrcopy.columns}"
-                )
+    df = self.get_dataframe()
+    df = _apply_pfilter_to_dataframe(df, pfilter)
+    if df.empty:
```

Comment:
You might not actually need to return the dataframe from this function, I would guess (but am not sure and have not tested!) it would edit it in-place. But it's still good to return it.